matrix-sync.c

```diff
@@ -291,5 +296,41 @@ void matrix_sync_parse(PurpleConnection *pc, JsonNode *body,
         g_list_free(room_ids);
     }
 
+    /* Handle d2d messages so we can create any e2e sessions needed
+     * We need to do this after we created rooms/conversations, but before
+     * we handle timeline events that we might need to decrypt.
+     */
+    JsonObject *to_device = matrix_json_object_get_object_member(rootObj,
+                               "to_device");
+    if (to_device) {
+        JsonArray *events = matrix_json_object_get_array_member(to_device,
+                                                               "events");
+        guint i = 0;
+        JsonNode *device_event;
+        while (device_event = matrix_json_array_get_element(events, i++),
+               device_event) {
+            JsonObject *event_obj = matrix_json_node_get_object(device_event);
+            const gchar *event_type;
+            event_type = matrix_json_object_get_string_member(event_obj,
+                                                               "type");
+            purple_debug_info("matrixprpl",  "to_device: Got %s from %s\n",
+                    event_type,
+                    matrix_json_object_get_string_member(event_obj, "sender"));
+            // TODO Handle some events
+        }
+    }
+
+    /* Now go round the rooms again getting the timeline events */
```

Comment:
hrm, this isn't ideal. Maybe it would be better to completely split `matrix_sync_room` in two (one which creates the room if necessary, and processes the state, and the other which just processes the timeline: I don't think it's necessary to process the state a second time).

Alternatively: you're going to have to deal with megolm keys turning up after the events they correspond to anyway, so once you have that you could just process the to_device events after the room events.

Either way, it's probably fine for now, in the interest of getting something landed.